### PR TITLE
[CC-24379] Add folder resource and folder CRUD functionality

### PR DESCRIPTION
**Pull request by cameronnunez**
This change introduces a new `Folder` resource and adds the relevant CRUD functionality.

### CODE REVIEW

**Comment by cameronnunez** [comment]
Note: this PR is blocked by [CC-24377](https://cockroachlabs.atlassian.net/browse/CC-24377) but would still appreciate reviews to get early iterations in if needed

File: internal/provider/folder_resource.go [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+	}
+
+	var (
+		folderObj *client.Folder
+		err       error
+	)
+
+	// Rename the folder if new name specified.
+	if !plan.Name.IsNull() && plan.Name != state.Name {
+		apiResp, _, err = r.provider.service.RenameFolder(ctx, &client.RenameFolderRequest{
```
Comment (erademacher) [inline internal/provider/folder_resource.go:185]: This isn't specific to this PR, but I'd much rather see a general-purpose `UpdateFolder` PATCH endpoint than an RPC-style `RenameFolder`. Gives us a lot more flexibility in the future.

Comment (cameronnunez) [reply internal/provider/folder_resource.go:185]: Thanks for the review! That decision is outside the scope of this PR but I can ensure that the team discusses this on the CCAPI PR. Currently that PR is still a work in progress

**Comment by cameronnunez** [comment]
making this PR temporarily **not ready for review**; the PR is currently in a state to allow others to test locally

**Comment by cameronnunez** [comment]
changes included as a part of another merged PR

**Pull request closed without merging**